core/src/main/scala/algebra/lattice/Logic.scala

```diff
@@ -0,0 +1,55 @@
+package algebra
+package lattice
+
+import scala.{specialized => sp}
+
+/**
+  * Logic models a logic generally.
+  * For intuitionistic logic see `Heyting`
+  * For fuzzy logic see `DeMorgan`
+  */
+trait Logic[@sp(Int, Long) A] extends Any { self =>
```

Comment:
Is this typeclass lawless?

Reply:
It is lawless at the moment. I am not sure what general laws common to both `Heyting` and `DeMorgan` I could use here.
If you have any input, I am interested to read it.

Reply:
They are both _bounded distributive lattices_.

One common property of the De Morgan involution and the pseudocomplement operation is that _a ≤ ¬¬a_. Not sure if it is terribly useful.

The question here is, what use of `Logic` do you imagine?

Reply:
I want to express a structure which is very much like a *bounded distributive lattice* with an extra unary operator.
I would like it not to use lattice operations (`meet, join`) but use instead common logic operators (`and, or, xor, not, imp,...`).

Do you think I should skip the `Logic` bit and come up with a name for a *bounded distributed lattice* with an extra unary operator ?
i.e. in the company code I am currently making open-source, `DeMorgan` is actually `InvolutiveBoundedLattice`

Reply:
> I want to express a structure which is very much like a bounded distributive lattice with an extra unary operator.

Surely you don't want any arbitrary unary operation. So let's try to formulate what properties it should satisfy in relation to the other operations.

> Do you think I should skip the `Logic` bit and come up with a name for a bounded distributed lattice with an extra unary operator ?

I am not concerned about the name, yet.

From [Wikipedia](https://en.wikipedia.org/wiki/Heyting_algebra#The_De_Morgan_laws_in_a_Heyting_algebra):

> One of the two De Morgan laws is satisfied in every Heyting algebra, namely
>
>     ¬(x ∨ y) = ¬x ∧ ¬y
>
> However, the other De Morgan law does not always hold. We have instead a weak de Morgan law:
>
>     ¬(x ∧ y) = ¬¬(¬x ∨ ¬y)